### PR TITLE
:boom: Minor breaking changes

### PR DESCRIPTION
**Pull request by lambdalisue**
#### [🐛 Fix fire timing of DenopsReady autocmd](https://github.com/vim-denops/denops.vim/commit/860918ecb54872ef422a47c9b998481f356f4b33) 

Previously, the `DenopsReady` autocmd was fired when the channel was connected. However, it is not the right timing because the denops itself may not be ready yet. This commit fixes the timing of the autocmd so that denops is ready when the autocmd is fired.

If this change cause issue, developers should use `denops#server#wait()` or `denops#server#wait_async()` function to ensure that functions are called AFTER denops got ready. For example

```vim
call denops#server#wait_async({ -> denops#plugin#load('...', '...') })
```

#### [💥 Read meta only once in initialization](https://github.com/vim-denops/denops.vim/commit/e91e0a1f9a1282135798c9768aeef4aced6e64a8) 

This is a minor but a breaking change. This change affects the `g:denops#debug` variable. After this commit, the variable must be configured prior to the denops initialization.

#### [💥 Synchronously call wait_async() callbacks](https://github.com/vim-denops/denops.vim/pull/300/commits/ac183c98b7e76cc69dfe41b5234bb87d9ae259d4) 

It seems `timer_start()` triks is no longer needed so we can call callbacks synchronously in both cases.

**Let me know** if this cause acutal issues (it might break lazy loading mech of denops that are not tested by myself.)

#### [💥 Remove g:denops#trace option](https://github.com/vim-denops/denops.vim/pull/300/commits/29b60ef0d4a44baf3fd84de482a447d3259fe4a6) 

The trace feature is not available when denops liberates from Worker and the feature obstructs the development of denops.

#### [💥 Remove DenopsPluginRegister:* autocmd](https://github.com/vim-denops/denops.vim/pull/300/commits/99f776334fa29b1226eb25e75907e97cf90e1d1c) 

We are planning to change internal registration mechanism and the autocmd is not suited for the new mechanism.

Quick search on GitHub shows that the autocmd is not used by anyones.

https://github.com/search?q=DenopsPluginRegister&type=code

#### [💥 Remove {options} from register, reload, discover](https://github.com/vim-denops/denops.vim/pull/300/commits/37bfbfdf0b7e7935184ffab6475774a087ceea26) 

We found that developers always use `mode: "skip"` and other options are not necessary. So we decided to remove `{options}` from functions and always use `mode: "skip"`.

Quick search revelaed that no one uses `reload` and `discover` with `{options}` but some (e.g. dein.vim) uses `register` with `{options}`.

That's why we introduced a new `load` function that replaces `register`.

Developers who need to support previous denops should use `load` whenever possible like:

```vim
try
  call denops#plugin#load(
        \ "denops-hello",
        \ "/path/to/denops-hello/main.ts",
        \)
catch /^Vim\%((\a\+)\)\=:E117:/
  " Fallback to `register` for backward compatibility
  call denops#plugin#register(
        \ "denops-hello",
        \ "/path/to/denops-hello/main.ts",
        \ {"mode": "skip"},
        \)
endtry
```

## Summary by CodeRabbit

- **Refactor**
	- Streamlined plugin registration and reloading processes.
	- Enhanced internal message handling and dispatching mechanisms.

- **New Features**
	- Implemented a more efficient signaling system within the server connection workflow.

- **Documentation**
	- Updated the requirement for setting the debug configuration prior to initialization.

- **Bug Fixes**
	- Addressed control flow issues in connection handling and service invocation.

### CODE REVIEW

**Comment by lambdalisue** [comment]
It seems tests are broken. I'll fix later

**Comment by lambdalisue** [comment]
Fixed

**Comment by Shougo** [comment]
The conflict exists.

**Comment by Shougo** [comment]
I will test it.

File: denops/@denops-private/host/nvim.ts [hunk 1]
```diff
@@ -33,6 +34,25 @@ export class Neovim implements Host {
     this.#session = new Session(reader, writer, {
       errorSerializer,
     });
+    this.#session.dispatcher = {
+      nvim_error_event(type, message) {
+        console.error(`nvim_error_event(${type})`, message);
+      },
+
+      void() {
+        return Promise.resolve();
```
[the remark this answers is not in the document]
Comment (lambdalisue) [reply denops/@denops-private/host/nvim.ts:43]: Error details should be shown so it's ok.

**Pull request merged**